### PR TITLE
Target Hasher rework

### PR DESCRIPTION
**Pull request by kenotron**
Should address both #639 & #640

This is a LARGE change, we are going to bump the lage version to another minor version due to the magnitude of this change.

1. TargetHasher has been moved from @lage-run/cache -> @lage-run/hasher
2. SimpleScheduler is now taking in a target hasher as a dep, and then instantiates its own cache provider
3. hrtime formatter is fixed with <1 second in process.hrtime diffs
4. tests are updated
5. added FileHasher that will grab git hashes for the repo on the first run, then writes to a hash manifest (looks up hash based on mtimeMs + size)
6. added PackageTree that will help get all the files in a micromatch pattern for packages - keeps indexed in folder tree structure for fast lookup - memoized so that lookups of the same package with the same patterns are faster
7. removed unused files

## Details

We have moved the target hasher work from doing its work inside workers to back in the main thread. This is needed because everything that was hashed could benefit from having previously known hashes to be looked up rather than calculated again from content reads. The shared cache of hashes means that all target hashes are accumulated as the job is run. The TargetHasher also keeps track of hashes of previously calculated target hashes as part of an input for the dependent targets. The target hashes are calculated now based on:

1. env globs
2. package dependency name@versions for external deps
3. package dependency names for internal deps
4. target inputs - defaults now to just the source controlled files in its own package folder, but it is flexible now to support a syntax like "^**/*" meaning the src files from the internal transitive dependency packages

## Mechanism of getting hashes from main thread from workers

When TargetHasher is moved to the main thread, we cannot naively simply calculate the hashes greedily or else we risk stalling the main thread from scheduling work for the worker pool. We have to use the workers themselves to request for the hash. The method used is via the `postMessage()` mechanism. When the worker is assigned a target to run, it will first request for the main thread to calculate the hash at that time. This is also done to respect the concurrency setting from lage's configurations.

### CODE REVIEW

File: packages/hasher/src/FileHasher.ts [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+  }
+
+  async getHashesFromGit() {
+    const { root } = this.options;
+    const fileHashes = await getPackageDeps(root);
+    const files = [...fileHashes.keys()];
+    const fileStats = stat(files, { cwd: root }) ?? {};
+
+    for (const [relativePath, fileStat] of Object.entries(fileStats)) {
+      if (fileHashes.has(relativePath)) {
```
Comment (dzearing) [inline packages/hasher/src/FileHasher.ts:37]: minor, but you can also just do something like:

```
const hash = fileHashes.get(relativePath);

if (hash) {
  const { mtime, size } = stat  this.#store[relativePath] = stat(files, ...);

  this.#store[relativePath] = { hash, size, mtime };
}
```

Comment (dzearing) [reply packages/hasher/src/FileHasher.ts:37]: this makes for less hash lookups and stat is only called if hash is available.

File: packages/hasher/src/FileHasher.ts [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+      });
+
+      inputStream.on("end", () => {
+        rl.close();
+        resolve();
+      });
+    });
+  }
+
+  writeManifest() {
```
Comment (dzearing) [inline packages/hasher/src/FileHasher.ts:78]: kinda odd that read is async but write is sync. (not a big deal, just was odd)

Comment (kenotron) [reply packages/hasher/src/FileHasher.ts:78]: Yeah, streams are weird for the read -> readline interface thing - needed to convert it to async. Writing in sync basically blocks me from doing too many async writes to that file. I am also using graceful-fs to make sure our watches would be okay if lots of concurrent requests came in. On another PR, I'd probably convert this to an async with a debouncer.

File: packages/cli/src/cache/isRunningFromCI.ts [hunk 3]
```diff
@@ -0,0 +1 @@
+export const isRunningFromCI = process.env.NODE_ENV !== "test" && (!!process.env.CI || !!process.env.TF_BUILD);
```
Comment (dzearing) [inline packages/cli/src/cache/isRunningFromCI.ts:1]: i can't find where this is used.

File: packages/cli/src/cache/createCacheProvider.ts [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+
+interface CreateCacheOptions {
+  cacheOptions?: CacheOptions;
+  logger: Logger;
+  root: string;
+  skipLocalCache: boolean;
+  cliArgs: string[];
+}
+
+export async function createCache(options: CreateCacheOptions) {
```
Comment (dzearing) [inline packages/cli/src/cache/createCacheProvider.ts:13]: rename file to `createCache` to align with export

File: packages/scheduler/src/WrappedTarget.ts [hunk 5]
```diff
@@ -146,33 +148,27 @@ export class WrappedTarget implements TargetRun {
...
-            cachedOutput.on("close", () => {
-              this.onSkipped(hash);
-              resolve();
-            });
-          });
+        const shouldShowCachedOutput = fs.existsSync(cachedOutputFile);
+        if (shouldShowCachedOutput) {
+          const cachedOutput = fs.readFileSync(cachedOutputFile, "utf8");
+          logger.verbose(">> Replaying cached output", { target });
+          logger.verbose(cachedOutput.trim(), { target });
```
Comment (dzearing) [inline packages/scheduler/src/WrappedTarget.ts:171]: Should this be `logger.stream`?

**Review by dzearing — APPROVED** [review]

File: packages/scheduler/src/WrappedTarget.ts (hunk 5 above)
Comment (kenotron) [reply packages/scheduler/src/WrappedTarget.ts:171]: I did the stream before - and the problem was that the close of the stream fought with the idea of being able to tell when the stream was consumed completely from the logger and ended up having summary printed before all the log streams were printed... at least that's ONE source of that mess!

**Pull request merged**